File: demo-src/example-board/demo-task-executor.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import base64
import hashlib
import json
import os
import subprocess
import sys
import tempfile
import time
import urllib.request
from typing import Any, Dict, Optional

_HERE = os.path.dirname(os.path.abspath(__file__))
_PYCLI_ROOT = os.path.normpath(os.path.join(_HERE, "..", "..", "pycli"))
if _PYCLI_ROOT not in sys.path:
    sys.path.insert(0, _PYCLI_ROOT)

from pylib.cli.storage_interface import parse_ref

MOCK_DB: Dict[str, Any] = {
    "quotes": {
        "quoteResponse": {
            "result": [
                {"symbol": "AAPL", "shortName": "Apple Inc.", "regularMarketPrice": 198.15, "regularMarketChange": 2.15, "regularMarketChangePercent": 1.10},
                {"symbol": "MSFT", "shortName": "Microsoft Corp.", "regularMarketPrice": 415.32, "regularMarketChange": -1.23, "regularMarketChangePercent": -0.30},
                {"symbol": "GOOGL", "shortName": "Alphabet Inc.", "regularMarketPrice": 174.89, "regularMarketChange": 0.89, "regularMarketChangePercent": 0.51},
                {"symbol": "TSLA", "shortName": "Tesla Inc.", "regularMarketPrice": 247.12, "regularMarketChange": 5.43, "regularMarketChangePercent": 2.25},
            ],
            "error": None,
        },
    },
}

CACHE_DIR = os.path.join(tempfile.gettempdir(), "demo-executor-cache")
DEFAULT_CACHE_TTL_MS = 60 * 60 * 1000


def _resolve_ref_to_path(ref: str) -> str:
    if ref.startswith("b64:"):
        parsed = parse_ref(ref)
        if parsed.get("kind") != "fs-path":
            raise ValueError(f"Unsupported ref kind for file IO: {parsed.get('kind')}")
        return str(parsed.get("value") or "")
    return ref


def _read_json_file(ref: str) -> Dict[str, Any]:
    path = _resolve_ref_to_path(ref)
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, dict):
        raise ValueError("Input JSON root must be an object")
    return data


def _write_json_file(ref: str, payload: Any) -> None:
    path = _resolve_ref_to_path(ref)
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(payload, f, indent=2, ensure_ascii=True)


def _write_err(err_ref: Optional[str], msg: str) -> None:
    if not err_ref:
        return
    try:
        path = _resolve_ref_to_path(err_ref)
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, "w", encoding="utf-8") as f:
            f.write(msg)
    except Exception:
        pass


def _interpolate(template: str, args: Dict[str, Any]) -> str:
    out = template
    for key, value in args.items():
        needle = "{{" + str(key) + "}}"
        if needle in out:
            out = out.replace(needle, value if isinstance(value, str) else json.dumps(value, ensure_ascii=True))
    return out


def _detect_kind(source_def: Dict[str, Any]) -> str:
    if "url" in source_def:
        return "url"
    if "url-list" in source_def:
        return "url-list"
    if "copilot" in source_def or "prompt_template" in source_def:
        return "copilot"
    if "workiq" in source_def:
        return "workiq"
    if "mock" in source_def:
        return "mock"
    raise ValueError("No recognised source kind")


def _cache_key(seed: str) -> str:
    return hashlib.sha1(seed.encode("utf-8")).hexdigest()


def _read_cache(key: str, ttl_ms: int) -> Any:
    path = os.path.join(CACHE_DIR, f"{key}.json")
    try:
        st = os.stat(path)
        age_ms = int(time.time() * 1000) - int(st.st_mtime * 1000)
        if age_ms > ttl_ms:
            return None
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return None


def _write_cache(key: str, value: Any) -> None:
    os.makedirs(CACHE_DIR, exist_ok=True)
    with open(os.path.join(CACHE_DIR, f"{key}.json"), "w", encoding="utf-8") as f:
        json.dump(value, f, ensure_ascii=True)


def _fetch_json(url: str, method: str, headers: Dict[str, str]) -> Any:
    req = urllib.request.Request(url=url, method=method.upper())
    for k, v in headers.items():
        req.add_header(str(k), str(v))
    with urllib.request.urlopen(req, timeout=10) as resp:
        raw = resp.read().decode("utf-8")
    return json.loads(raw)


def _resolve_tickers_arg(source_def: Dict[str, Any], fetch_args: Dict[str, Any]) -> None:
    tickers_from = source_def.get("tickersFrom")
    if not isinstance(tickers_from, str) or "." not in tickers_from:
        return
    ref_key, field_name = tickers_from.split(".", 1)
    arr = (source_def.get("_projections") or {}).get(ref_key)
    if not isinstance(arr, list):
        return
    vals = [row.get(field_name) for row in arr if isinstance(row, dict) and row.get(field_name)]
    if vals:
        fetch_args["tickers"] = ",".join(str(v) for v in vals)


def _execute_url(source_def: Dict[str, Any]) -> Any:
    cfg = source_def.get("url")
    if not isinstance(cfg, dict):
        raise ValueError("url source requires object config")
    method = str(cfg.get("method") or "GET").upper()
    headers = dict(cfg.get("headers") or {})
    cache_timeout = cfg.get("cacheTimeout")
    ttl_ms = int(cache_timeout * 1000) if isinstance(cache_timeout, (int, float)) else DEFAULT_CACHE_TTL_MS
    fetch_args = dict(cfg.get("args") or {})

    _resolve_tickers_arg(source_def, fetch_args)
    context = {}
    context.update(source_def.get("_projections") or {})
    context.update(fetch_args)
    url_tpl = cfg.get("url")
    if not isinstance(url_tpl, str):
        raise ValueError("url source missing url template")
    url = _interpolate(url_tpl, context)

    key = _cache_key(f"url:{method}:{url}")
    cached = _read_cache(key, ttl_ms)
    if cached is not None:
        return cached

    data = _fetch_json(url, method, headers)
    _write_cache(key, data)
    return data


def _execute_url_list(source_def: Dict[str, Any]) -> Any:
    cfg = source_def.get("url-list")
    if not isinstance(cfg, dict):
        raise ValueError("url-list source requires object config")
    method = str(cfg.get("method") or "GET").upper()
    headers = dict(cfg.get("headers") or {})
    cache_timeout = cfg.get("cacheTimeout")
    ttl_ms = int(cache_timeout * 1000) if isinstance(cache_timeout, (int, float)) else DEFAULT_CACHE_TTL_MS
    url_list = (source_def.get("_projections") or {}).get("url_list")
    if not isinstance(url_list, list) or not url_list:
        raise ValueError("url-list source requires _projections.url_list as non-empty array")

    out = []
    for item in url_list:
        url = str(item)
        key = _cache_key(f"url-list:{method}:{url}")
        cached = _read_cache(key, ttl_ms)
        if cached is not None:
            out.append(cached)
            continue
        data = _fetch_json(url, method, headers)
        _write_cache(key, data)
        out.append(data)
    return out


def _execute_copilot(source_def: Dict[str, Any], out_ref: str) -> Any:
    cfg = source_def.get("copilot") if isinstance(source_def.get("copilot"), dict) else {}
    template = cfg.get("prompt_template") or source_def.get("prompt_template")
    if not isinstance(template, str) or not template:
        raise ValueError("copilot source missing prompt_template")

    args = {}
    args.update(source_def.get("_projections") or {})
    args.update(cfg.get("args") or source_def.get("args") or {})
    prompt = _interpolate(template, args)

    wrapper = os.path.join(_HERE, "scripts", "copilot_wrapper.bat")
    if os.name != "nt" or not os.path.isfile(wrapper):
        raise RuntimeError("copilot wrapper is unavailable in this environment")

    out_file = _resolve_ref_to_path(out_ref) + ".copilot.json"
    prompt_file = out_file + ".prompt.txt"
    with open(prompt_file, "w", encoding="utf-8") as f:
        f.write(prompt)
    try:
        subprocess.run(
            [
                "cmd.exe", "/d", "/c", wrapper,
                out_file,
                os.path.join(tempfile.gettempdir(), "demo-task-executor-copilot"),
                os.getcwd(),
                "@" + prompt_file,
                "json",
                str(source_def.get("bindTo") or "executor"),
                "",
                "",
            ],
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        with open(out_file, "r", encoding="utf-8") as f:
            return json.load(f)
    finally:
        for p in (prompt_file, out_file):
            try:
                os.remove(p)
            except OSError:
                pass


def _execute_workiq(source_def: Dict[str, Any]) -> Any:
    raise RuntimeError("workiq source is not available in Python-only standalone")


def _execute_mock(source_def: Dict[str, Any]) -> Any:
    key = source_def.get("mock")
    if not isinstance(key, str) or not key:
        raise ValueError("mock source requires a string key")
    if key not in MOCK_DB:
        raise ValueError(f"mock key not found: {key}")
    return MOCK_DB[key]


def _run_source_fetch(in_ref: str, out_ref: str, err_ref: Optional[str]) -> int:
    try:
        source_def = _read_json_file(in_ref)
        kind = _detect_kind(source_def)
        if kind == "url":
            result_value = _execute_url(source_def)
        elif kind == "url-list":
            result_value = _execute_url_list(source_def)
        elif kind == "copilot":
            result_value = _execute_copilot(source_def, out_ref)
        elif kind == "workiq":
            result_value = _execute_workiq(source_def)
        else:
            result_value = _execute_mock(source_def)
        _write_json_file(out_ref, result_value)
        return 0
    except Exception as err:
        _write_err(err_ref, str(err))
        print(f"[demo-task-executor.py] {err}", file=sys.stderr)
        return 1


def _describe_capabilities() -> int:
    payload = {
        "ok": True,
        "kinds": ["url", "url-list", "copilot", "workiq", "mock"],
        "python_only": True,
    }
    print(json.dumps(payload, ensure_ascii=True))
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description="Python demo task executor")
    parser.add_argument("subcommand", choices=["run-source-fetch", "describe-capabilities"])
    parser.add_argument("--in-ref", dest="in_ref")
    parser.add_argument("--out-ref", dest="out_ref")
    parser.add_argument("--err-ref", dest="err_ref")
    parser.add_argument("--extra", dest="extra", required=False)
    args = parser.parse_args()

    if args.subcommand == "describe-capabilities":
        return _describe_capabilities()
    if not args.in_ref or not args.out_ref:
        print("run-source-fetch requires --in-ref and --out-ref", file=sys.stderr)
        return 2
    return _run_source_fetch(args.in_ref, args.out_ref, args.err_ref)


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/board/demo-task-executor.js
#!/usr/bin/env node

/**
 * demo-task-executor.js — Simple mock source executor for example-board.
 *
 * Subcommands:
 *   run-source-fetch        — fetch data for one source entry
 *   describe-capabilities   — print supported source kinds + schemas to stdout (JSON)
 *
 * CLI args:
 *   --in    <source.json>   Required. Path to a temp JSON file containing the source definition.
 *   --out   <result.json>   Required. Path where this executor must write its JSON result.
 *   --err   <error.txt>     Optional. Path where this executor writes an error message on failure.
 *   --extra <base64json>    Optional. Base64-encoded JSON with board topology context
 *                           (baked into .task-executor at board init time, passed blindly by the CLI).
 *
 * --in payload (source definition):
 *   {
 *     "bindTo":  "token_name",
 *     "outputFile": "relative/path.json",
 *     "cwd":     "<card directory>",           // injected by CLI
 *     "boardDir":"<board runtime directory>",   // injected by CLI
 *     "_projections":   { "refKey": <resolvedValue> }, // named projections from card_data/requires,
 *                                               // declared in source_defs[].projections and resolved
 *                                               // by the engine before invoking the executor
 *     // ...plus any custom fields authored on the source entry (bindTo, outputFile, projections, etc.)
 *   }
 *
 * --extra (decoded):
 *   {
 *     "boardSetupRoot":   "<abs path>",        // board root (parent of runtime/, surface/, runtime-out/)
 *     "boardId":          "<board id>",        // e.g. "default"
 *     "boardRuntimeDir":  "<relative>",        // e.g. "runtime"
 *     "runtimeStatusDir": "<relative>",        // e.g. "runtime-out"
 *     "cardsDir":         "<relative>",        // e.g. "surface/tmp-cards"
 *     "serverUrl":        "<base url>",        // optional; e.g. "http://127.0.0.1:7799"
 *     "boardLiveCardsCliJs":"<abs path>",      // optional; path to board-live-cards-cli.js
 *     "stepMachineCliPath":"<abs path>"        // optional; path to step-machine-cli.js
 *   }
 *
 * Supported source kinds (based on custom fields in --in):
 *   - { mock: "key" }              → look up key in MOCK_DB (hardcoded below)
 *   - { copilot: { prompt_template, args? } }  → call Copilot CLI with interpolated prompt
 *   - { prompt_template: "..." }   → shorthand copilot call (top-level template)
 *   - { workiq: { query_template, args? } }   → call WorkIQ (M365 Copilot) with interpolated query
 *   - { "url": { url, method?, headers?, args?, cacheTimeout? }, tickersFrom? }
 *       → single URL fetch via curl with {{key}} interpolation from _projections
 *   - { "url-list": { method?, headers?, cacheTimeout? } }
 *       → fan-out over _projections.url_list (string[]); returns array of responses.
 *         Build url_list in projections: e.g. `requires.holdings.ticker.('https://host/' & $ & '?q=1')`
 *     Prefer url-list for multi-URL fan-out sources.
 *   A real executor can also handle: graphapi, teams, mail, incidentdb, script, etc.
 *
 * url / url-list notes:
 *   - Results cached in os.tmpdir()/demo-executor-cache/ per URL (default 1 hour, override via cacheTimeout)
 */

import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { execFileSync } from 'node:child_process';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { parseRef, blobStorageForRef, reportComplete, reportFailed } from 'yaml-flow/board-worker-adapter';
import { loadStepFlow, createStepMachine } from '../../lib/step-machine/index.js';
import { MemoryStore } from '../../lib/stores/memory.js';
import { buildStepHandlersForFlow } from '../../lib/step-machine-public/index.js';
import { invokeRefSync } from '../../cli/node/execution-adapter.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SOURCE_DEF_FLOWS_FILE = path.join(__dirname, 'source_def_flows.json');

// ---------------------------------------------------------------------------
// Mock data — used when a source has { mock: "key" }.
// Edit these values to change the demo data without needing a mock.db file.
// ---------------------------------------------------------------------------
const MOCK_DB = {
  quotes: {
    quoteResponse: {
      result: [
        { symbol: 'AAPL',  shortName: 'Apple Inc.',      regularMarketPrice: 198.15, regularMarketChange:  2.15, regularMarketChangePercent:  1.10 },
        { symbol: 'MSFT',  shortName: 'Microsoft Corp.', regularMarketPrice: 415.32, regularMarketChange: -1.23, regularMarketChangePercent: -0.30 },
        { symbol: 'GOOGL', shortName: 'Alphabet Inc.',   regularMarketPrice: 174.89, regularMarketChange:  0.89, regularMarketChangePercent:  0.51 },
        { symbol: 'TSLA',  shortName: 'Tesla Inc.',      regularMarketPrice: 247.12, regularMarketChange:  5.43, regularMarketChangePercent:  2.25 },
      ],
      error: null,
    },
  },
};

function readJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
}

function interpolatePrompt(template, args) {
  return String(template).replace(/\{\{\s*([a-zA-Z0-9_.-]+)\s*\}\}/g, (_, key) => {
    const v = args?.[key];
    if (v === undefined) return '';
    if (typeof v === 'string') return v;
    return JSON.stringify(v);
  });
}

// Reusable prompt fragments available to all copilot source templates.
// Source definitions can interpolate them with {{view_kind_guidance}} and {{card_layout_guidance}}.
const COPILOT_PROMPT_CONTEXT = {
  view_kind_guidance: [
    'VIEW KIND GUIDANCE (for dynamic ref rendering):',
    '- Return a _view object whenever your output data is meant for a ref element.',
    '- Allowed _view.kind values only: table, editable-table, chart, metric, list, badge, text, narrative, markdown, form, filter, todo, alert.',
    '- If uncertain, use "table".',
    '- For array rows that users should edit, prefer "editable-table" and set _view.data.writeTo to a card_data path.',
    '- For chart, set _view.data.chartType and _view.data.columns with [labelField, valueField].',
    '- Keep _view.data minimal and valid JSON (no comments, no trailing text).',
  ].join('\n'),
  card_layout_guidance: [
    'CARD LAYOUT GUIDANCE:',
    '- Prefer compact outputs that fit a card: one primary structure plus concise rationale text.',
    '- Avoid repeating values already present in upstream inputs.',
    '- If you produce both machine-readable and human-readable content, keep machine-readable fields top-level and concise prose in a separate field.',
  ].join('\n'),
};

function resolveCopilotPrompt(sourceDef) {
  const cfg = sourceDef?.copilot && typeof sourceDef.copilot === 'object' ? sourceDef.copilot : {};
  const template = cfg.prompt_template ?? sourceDef.prompt_template;
  const args = cfg.args ?? cfg.prompt_args ?? sourceDef.prompt_args ?? sourceDef.args ?? {};
  
  // Merge _projections into template interpolation context.
  // _projections contains the named data projections declared in source_defs[].projections,
  // evaluated by the engine from card_data/requires before invoking this executor.
  // Explicit args defined on the source take highest precedence.
  const interpolationContext = {
    ...COPILOT_PROMPT_CONTEXT,
    ...sourceDef._projections,
    ...args,
  };
  
  if (!template || typeof template !== 'string') return null;
  return interpolatePrompt(template, interpolationContext);
}

/**
 * Run a copilot prompt via copilot_wrapper.bat (Windows only).
 *
 * The wrapper handles:
 *   - Session management (--resume UUID for multi-turn continuity)
 *   - Noise/footer stripping (via copilot_wrapper_helper.ps1)
 *   - JSON mode extraction with optional result_shape key matching
 *   - Agentic retry: if the first response isn't valid JSON, the wrapper calls
 *     copilot again in the same session with a correction prompt, then re-extracts.
 *
 * @param {string} prompt         - interpolated prompt string
 * @param {object} sourceDef      - source definition (may contain copilot.result_shape)
 * @param {string} wrapperOutFile - path the wrapper writes its JSON output to
 * @param {string} sessionDir     - persistent dir for session UUID (enables --resume)
 * @param {string} cwd            - working directory for copilot (boardSetupRoot)
 * @returns {unknown} parsed JSON result value
 */
function runCopilotViaWrapper(prompt, sourceDef, wrapperOutFile, sessionDir, cwd) {
  const wrapperPath = path.join(__dirname, 'scripts', 'copilot_wrapper.bat');

  const promptFile = wrapperOutFile + '.prompt.txt';
  fs.writeFileSync(promptFile, prompt, 'utf-8');

  // Optional result_shape_file: top-level keys the response JSON must contain.
  // Sourced from sourceDef.copilot.result_shape or sourceDef.result_shape.
  let shapeFile = '';
  const shape = sourceDef?.copilot?.result_shape ?? sourceDef?.result_shape;
  if (shape && typeof shape === 'object') {
    shapeFile = wrapperOutFile + '.shape.json';
    fs.writeFileSync(shapeFile, JSON.stringify(shape), 'utf-8');
  }

  fs.mkdirSync(sessionDir, { recursive: true });

  try {
    execFileSync('cmd.exe', [
      '/d', '/c',
      wrapperPath,
      wrapperOutFile,                    // OUTPUT_FILE
      sessionDir,                        // SESSION_DIR
      cwd || process.cwd(),             // WORKING_DIR
      '@' + promptFile,                 // REQUEST_OR_FILE (@ prefix = file path)
      'json',                           // RESULT_TYPE — wrapper extracts JSON + retries
      sourceDef.bindTo || 'executor',   // AGENT_NAME (for log file naming)
      '',                               // MODEL (empty = wrapper default)
      shapeFile,                        // RESULT_SHAPE_FILE (empty = accept any JSON)
    ], {
      encoding: 'utf-8',
      stdio: ['ignore', 'pipe', 'pipe'],
      maxBuffer: 10 * 1024 * 1024,
      windowsHide: true,
    });
  } finally {
    try { fs.unlinkSync(promptFile); } catch {}
    if (shapeFile) { try { fs.unlinkSync(shapeFile); } catch {} }
  }

  return JSON.parse(fs.readFileSync(wrapperOutFile, 'utf-8').replace(/^\uFEFF/, ''));
}

function fail(msg, errFile) {
  if (errFile) {
    try {
      fs.writeFileSync(errFile, msg);
    } catch {}
  }
  console.error(`[demo-task-executor] ${msg}`);
  process.exit(1);
}

function loadSourceDefFlowsConfig() {
  try {
    return readJson(SOURCE_DEF_FLOWS_FILE);
  } catch (err) {
    fail(`Cannot read source flow registry at ${SOURCE_DEF_FLOWS_FILE}: ${String(err && err.message || err)}`);
  }
}

function matchesDetectRule(sourceDef, detect) {
  if (!detect || typeof detect !== 'object') return false;
  if (typeof detect.field === 'string') {
    return sourceDef[detect.field] !== undefined;
  }
  if (Array.isArray(detect.anyOfFields)) {
    return detect.anyOfFields.some((field) => sourceDef[field] !== undefined);
  }
  return false;
}

function resolveSourceKind(sourceDef, registry) {
  const kinds = registry?.kinds && typeof registry.kinds === 'object' ? registry.kinds : {};
  const order = Array.isArray(registry?.resolveOrder) ? registry.resolveOrder : Object.keys(kinds);
  const matched = [];
  for (const kind of order) {
    const spec = kinds[kind];
    if (!spec) continue;
    if (matchesDetectRule(sourceDef, spec.detect)) {
      matched.push(kind);
    }
  }

  if (matched.length === 0) {
    const knownKinds = Object.keys(kinds);
    throw new Error(`No recognised source kind. Known kinds: ${knownKinds.join(', ')}`);
  }
  if (matched.length > 1) {
    throw new Error(`Multiple source kinds specified: [${matched.join(', ')}]. Use exactly one.`);
  }
  return matched[0];
}

async function executeStepMachineSourceFlow(context) {
  const { kind, registry } = context;
  const spec = registry?.kinds?.[kind];
  if (!spec) {
    throw new Error(`Missing flow registration for kind: ${kind}`);
  }

  const flowRef = spec.flow;
  if (typeof flowRef !== 'string' || flowRef.length === 0) {
    throw new Error(`Invalid or missing flow for kind: ${kind}`);
  }

  const flowPath = path.resolve(__dirname, flowRef);
  const flow = await loadStepFlow(flowPath);

  const invokeHttpRef = async (ref, args) => {
    const rawUrl = typeof ref.whatToRun === 'object' ? ref.whatToRun.value : parseRef(ref.whatToRun).value;

    const base = String(args?.extra?.serverUrl || 'http://127.0.0.1:7799').replace(/\/$/, '');
    const resolvedUrl = /^https?:\/\//i.test(rawUrl)
      ? rawUrl
      : `${base}${rawUrl.startsWith('/') ? '' : '/'}${rawUrl}`;

    let body = args;
    const workiqCfg = args?.sourceDef?.workiq;
    if (workiqCfg && typeof workiqCfg === 'object' && typeof workiqCfg.query_template === 'string') {
      const interpolationContext = {
        ...(args?.sourceDef?._projections || {}),
        ...(workiqCfg.args || {}),
      };
      body = {
        query: interpolatePrompt(workiqCfg.query_template, interpolationContext),
      };
    }

    const method = ref.howToRun === 'http:get' ? 'GET' : 'POST';
    const response = await fetch(resolvedUrl, {
      method,
      headers: { 'Content-Type': 'application/json' },
      ...(method === 'POST' ? { body: JSON.stringify(body) } : {}),
    });

    const text = await response.text();
    let parsed;
    try {
      parsed = text ? JSON.parse(text) : {};
    } catch {
      parsed = { response: text };
    }

    if (!response.ok) {
      const msg = typeof parsed?.error === 'string' ? parsed.error : `HTTP ${response.status}`;
      return { result: 'failure', data: { error: msg }, error: msg };
    }

    if (typeof parsed?.error === 'string') {
      return { result: 'failure', data: { error: parsed.error }, error: parsed.error };
    }

    return {
      result: 'success',
      data: {
        resultValue: Object.prototype.hasOwnProperty.call(parsed, 'response') ? parsed.response : parsed,
      },
    };
  };

  const invoke = async (ref, args) => {
    if (ref.howToRun === 'http:post' || ref.howToRun === 'http:get') {
      return invokeHttpRef(ref, args);
    }
    if (ref.howToRun === 'demo-local-module') {
      const whatValue = typeof ref.whatToRun === 'object' ? ref.whatToRun.value : parseRef(ref.whatToRun).value;
      const modulePath = path.resolve(__dirname, whatValue);
      const mod = await import(pathToFileURL(modulePath).href);
      if (typeof mod.execute !== 'function') {
        throw new Error(`Flow module ${JSON.stringify(ref.whatToRun)} must export execute(context)`);
      }
      return mod.execute(args);
    }
    return invokeRefSync(ref, args, { cliDir: __dirname, cwd: process.cwd() });
  };

  const handlers = buildStepHandlersForFlow(flow, { invoke });
  const machine = createStepMachine(flow, handlers, { store: new MemoryStore() });
  const run = await machine.run({
    ...context,
    promptContext: COPILOT_PROMPT_CONTEXT,
    executorDir: __dirname,
  });

  if (run.status !== 'completed') {
    const reason = run.error?.message ?? run.intent ?? run.status;
    throw new Error(`flow execution failed: ${reason}`);
  }

  if (run.intent !== 'success') {
    const reason = typeof run.data?.error === 'string' ? run.data.error : `flow returned intent: ${run.intent}`;
    throw new Error(reason);
  }

  return {
    resultValue: run.data?.resultValue,
    wroteOutputDirectly: !!run.data?.wroteOutputDirectly,
  };
}

async function runSourceFetchSubcommand(argv) {
  const inIdx = argv.indexOf('--in-ref');
  const outIdx = argv.indexOf('--out-ref');
  const errIdx = argv.indexOf('--err-ref');
  const extraIdx = argv.indexOf('--extra');
  const inRefStr  = inIdx  !== -1 ? argv[inIdx + 1]  : undefined;
  const outRefStr = outIdx !== -1 ? argv[outIdx + 1] : undefined;
  const errRefStr = errIdx !== -1 ? argv[errIdx + 1] : undefined;
  const extraB64  = extraIdx !== -1 ? argv[extraIdx + 1] : undefined;

  let extra = {};
  if (extraB64) {
    try { extra = JSON.parse(Buffer.from(extraB64, 'base64').toString('utf-8')); }
    catch { console.warn('[demo-task-executor] bad --extra base64, ignoring'); }
  }

  if (!inRefStr || !outRefStr) {
    fail('Usage: run-source-fetch --in-ref <b64:<base64url(json)>> --out-ref <b64:<base64url(json)>> [--err-ref <b64:<base64url(json)>>]');
  }

  let inRef, outRef, errRef;
  try {
    inRef  = parseRef(inRefStr);
    outRef = parseRef(outRefStr);
    if (errRefStr) errRef = parseRef(errRefStr);
  } catch (e) {
    fail(`invalid ref argument: ${e.message}`);
  }

  const inStorage  = blobStorageForRef(inRef);
  const outStorage = blobStorageForRef(outRef);
  const errStorage = errRef ? blobStorageForRef(errRef) : undefined;

  // Local error reporter — writes to errStorage and calls back to board if callback present.
  const failRef = (msg, callback) => {
    if (errStorage && errRef) { try { errStorage.write(errRef.value, msg); } catch {} }
    console.error(`[demo-task-executor] ${msg}`);
    if (callback) { try { reportFailed(callback, msg); } catch {} }
    process.exit(1);
  };

  const rawIn = inStorage.read(inRef.value);
  if (rawIn === null) {
    failRef(`Input not found: ${inRefStr}`);
  }

  // Payload may be { source_def, callback } (new protocol) or raw source def (legacy).
  let envelope;
  try {
    envelope = JSON.parse(rawIn);
  } catch (err) {
    failRef(`Cannot parse input: ${String(err && err.message || err)}`);
  }

  const callback = envelope.source_def ? envelope.callback : undefined;
  let sourceDef;
  try {
    sourceDef = envelope.source_def ?? envelope;
  } catch (err) {
    failRef(`Cannot resolve source_def: ${String(err && err.message || err)}`, callback);
  }

  const registry = loadSourceDefFlowsConfig();
  let kind;
  try {
    kind = resolveSourceKind(sourceDef, registry);
  } catch (err) {
    failRef(String(err && err.message || err), callback);
  }

  let flowResult;
  try {
    flowResult = await executeStepMachineSourceFlow({
      kind,
      registry,
      sourceDef,
      extra,
      inRef,
      outRef,
      errRef,
      mockDb: MOCK_DB,
    });
  } catch (err) {
    const detail = (err && (err.stderr || err.stdout)) ? `\n${err.stderr || err.stdout}`.trimEnd() : '';
    failRef(`${kind} invocation failed: ${String(err && err.message || err)}${detail}`, callback);
  }

  if (!flowResult?.wroteOutputDirectly) {
    try {
      outStorage.write(outRef.value, JSON.stringify(flowResult?.resultValue, null, 2));
    } catch (err) {
      failRef(`Cannot write output: ${String(err && err.message || err)}`, callback);
    }
  }

  if (callback) {
    try {
      reportComplete(callback, outRef);
    } catch (err) {
      console.error(`[demo-task-executor] reportComplete failed: ${String(err && err.message || err)}`);
      process.exit(1);
    }
  }

}

// ---------------------------------------------------------------------------
// validate-source-def — structural validation of a source definition
// ---------------------------------------------------------------------------
function validateSourceDefSubcommand(argv) {
  const inIdx = argv.indexOf('--in');
  const inFile = inIdx !== -1 ? argv[inIdx + 1] : undefined;

  if (!inFile) {
    console.error('[demo-task-executor] Usage: validate-source-def --in <source.json>');
    process.exit(1);
  }

  if (!fs.existsSync(inFile)) {
    console.log(JSON.stringify({ ok: false, errors: [`Input file not found: ${inFile}`] }));
    process.exit(1);
  }

  let sourceDef;
  try {
    sourceDef = readJson(inFile);
  } catch (err) {
    console.log(JSON.stringify({ ok: false, errors: [`Cannot parse source file: ${err && err.message || err}`] }));
    process.exit(1);
  }

  const errors = [];
  const registry = loadSourceDefFlowsConfig();

  let kind = '';
  try {
    kind = resolveSourceKind(sourceDef, registry);
  } catch (err) {
    errors.push(String(err && err.message || err));
  }

  // Data-driven validation: rules come from source_def_flows.json "validate" array
  if (kind) {
    const spec = registry.kinds?.[kind];
    const rules = Array.isArray(spec?.validate) ? spec.validate : [];
    for (const rule of rules) {
      if (rule.condition === 'copilot-or-prompt') {
        // Special condition: copilot object OR top-level prompt_template string
        const hasCopilotObj = typeof sourceDef.copilot === 'object';
        const hasTopLevelTemplate = typeof sourceDef.prompt_template === 'string';
        const hasNestedTemplate = hasCopilotObj && typeof sourceDef.copilot.prompt_template === 'string';
        if (!hasCopilotObj && !hasTopLevelTemplate) {
          errors.push(rule.message);
        } else if (hasCopilotObj && !hasNestedTemplate && !hasTopLevelTemplate) {
          errors.push('copilot.prompt_template is required (or use top-level prompt_template).');
        }
      } else if (rule.field) {
        // Dot-path field check: e.g. "url.url" → sourceDef.url.url
        const parts = rule.field.split('.');
        let val = sourceDef;
        for (const p of parts) { val = val != null ? val[p] : undefined; }
        if (val === undefined || val === null || typeof val !== rule.type) {
          errors.push(rule.message);
        }
      }
    }
  }

  const result = { ok: errors.length === 0, errors };
  console.log(JSON.stringify(result));
  process.exit(errors.length === 0 ? 0 : 1);
}

// ---------------------------------------------------------------------------
// describe-capabilities — introspection metadata for this executor
// Entirely derived from source_def_flows.json registry.
// ---------------------------------------------------------------------------
function describeCapabilities() {
  const registry = loadSourceDefFlowsConfig();
  const sourceKinds = Object.fromEntries(
    Object.entries(registry?.kinds || {}).map(([kind, spec]) => {
      const manifest = spec?.manifest && typeof spec.manifest === 'object' ? spec.manifest : {};
      return [kind, manifest];
    }),
  );
  const capabilities = {
    version: registry.version || '1.0',
    executor: registry.executor || 'demo-task-executor',
    subcommands: registry.subcommands || [],
    sourceKinds,
    extraSchema: registry.extraSchema || {},
  };
  console.log(JSON.stringify(capabilities, null, 2));
}

async function main() {
  const sub = process.argv[2];
  if (sub === 'run-source-fetch') {
    await runSourceFetchSubcommand(process.argv.slice(3));
    return;
  }
  if (sub === 'describe' || sub === 'describe-capabilities') {
    describeCapabilities();
    return;
  }
  if (sub === 'validate-source-def') {
    validateSourceDefSubcommand(process.argv.slice(3));
    return;
  }

  console.warn(`[demo-task-executor] Unknown subcommand: ${sub}`);
  process.exit(0);
}

main().catch(err => {
  console.error(`[demo-task-executor] fatal: ${err && err.message || err}`);
  process.exit(1);
});
